Sources/LeafKit/DocumentLoader.swift

```diff
@@ -0,0 +1,204 @@
+        for syntax in ast {
+            switch syntax {
+            case .extend, .export:
+                // all extentions MUST be resolved
+                throw "unresolved ast"
+            default:
+                continue
+            }
+        }
+        
+        self.name = name
+        self.ast = ast
+    }
+}
+
+final class DocumentLoader {
+    private var fileAccess: FileAccessProtocol
+    private var resolved: [String: ResolvedDocument] = [:]
+    private var unresolved: [String: UnresolvedDocument] = [:]
+    
+    init(_ access: FileAccessProtocol = FileAccessor()) {
+        self.fileAccess = access
+    }
+    
+    /// insert raw documents into the loader manually
+    @discardableResult
+    func insert(name: String, raw: String) throws -> UnresolvedDocument {
+        var buffer = ByteBufferAllocator().buffer(capacity: 0)
+        buffer.writeString(raw)
+        return try insert(name: name, raw: buffer)
+    }
+    
+    /// insert raw documents into the loader manually
+    @discardableResult
+    func insert(name: String, raw: ByteBuffer) throws -> UnresolvedDocument {
+        let document = try parse(name: name, raw: raw)
+        unresolved[name] = document
+        return document
+    }
+    
+    // insert raw documents into the loader manually
+    @discardableResult
+    func insert(name: String, raw: [Syntax]) -> UnresolvedDocument {
+        let document = UnresolvedDocument(name: name, raw: raw)
+        unresolved[name] = document
+        return document
+    }
+    
+    /// parse a raw body into an unresolved document
+    private func parse(name: String, raw: ByteBuffer) throws -> UnresolvedDocument {
+        var lexer = LeafLexer(template: raw)
+        let tokens = try lexer.lex()
+        var parser = LeafParser(tokens: tokens)
+        let syntax = try parser.parse()
+        return UnresolvedDocument(name: name, raw: syntax)
+    }
+    
+    /// Loads and resolves a document
+    
+    func load(_ name: String) throws -> ResolvedDocument {
+        if let cached = resolved[name] { return cached }
+        if let cached = unresolved[name] { return try resolve(cached) }
+        
+        // load raw from file
+        let buffer = try fileAccess.load(name: name)
+        let new = try parse(name: name, raw: buffer)
+        return try resolve(new)
+    }
+    
+    /// an individual object resolution
+    /// could probably be optimized
+    private func resolve(_ doc: UnresolvedDocument) throws -> ResolvedDocument {
+        unresolved[doc.name] = nil
+        try resolve([doc])
+        guard let value = resolved[doc.name] else { throw "unable to resolve \(doc)" }
+        return value
+    }
+    
+    // we're gonna be real lazy about this stop
+    // as opposed to trying to prioritize
+    // just keep checking what we can compile
+    // and if we can't, stick it in the back of
+    // the array and try again later
+    private func resolve(_ raw: [UnresolvedDocument]) throws {
```

Comment:
I'm not sure about the extend resolution being coupled with the document loader. I think things could be a lot simpler if we had separate steps and types for this. I like `DocumentLoader` as the view cache (I would even rename it to be something like `ViewCache`). But there should be a separate type like `ExtendResolver`:

```swift
struct ExtendResolver {
    var ast: [Syntax]
    var renderer: LeafRenderer
    func resolve() throws -> EventLoopFuture<[Syntax]> { ... }
}
```

`LeafRenderer` would invoke `ExtendResolver` after `LeafParser` but _before_ `LeafSerializer`. The `ExtendResolver` would use `LeafRenderer` (or maybe a delegate protocol?) to resolve any nested dependencies, recursing as needed. Finally, `LeafRenderer` would cache this result so that all further renders of the same file only need to serialize.

Reply:
let's talk about this, I had something similar originally.. actually, even called "ExtendResolver", it was moved around for some reasons